system-addon/lib/TopSitesFeed.jsm

```diff
@@ -259,6 +266,18 @@ this.TopSitesFeed = class TopSitesFeed {
     this._broadcastPinnedSitesUpdated();
   }
 
+  /**
+   * Handle a Top Site drop into new slot.
+   */
+  drop(action) {
+    const {site, index} = action.data;
+
+    // Dropping a top site pins it in the slot it was dropped, pushing over any link already
+    // pinned in the slot (unless it's the last slot, then it replaces).
+    this._insertPin(site, index);
+    this._broadcastPinnedSitesUpdated();
```

Comment:
I was thinking this looked familiar… ;) Perhaps we should have whatever would have called `add` triggering `_insertPin(site, 0)` to dispatch "DROP" although maybe the action should be more generically "TOP_SITES_INSERT" or just reuse "TOP_SITES_ADD" but have it take an `index` where content will either dispatch with a hardcoded 0 for the existing add behavior (or have a default index of 0?) and the dynamic index for the drop behavior.

Reply:
Ah, good point. No need to create new event type. Will fix